src/gbinder_fmq.c

```diff
@@ -0,0 +1,851 @@
+
+#include <gutil_intarray.h>
+#include <gutil_macros.h>
+
+#include <fcntl.h>
+#include <linux/futex.h>
+#include <stdatomic.h>
+#include <stdint.h>
+#include <sys/mman.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+/* Private */
+
+/* Grantor data positions */
+enum {
+    READ_PTR_POS = 0,
+    WRITE_PTR_POS,
+    DATA_PTR_POS,
+    EVENT_FLAG_PTR_POS
+};
+
+/* Wait bits for blocking read/write */
+enum {
+    FMQ_NOT_FULL = 0x01,
+    FMQ_NOT_EMPTY = 0x02
+};
+
+typedef struct gbinder_fmq_priv {
+    GBinderFmq pub;
+    guint8* ring;
+    _Atomic guint64* read_ptr;
+    _Atomic guint64* write_ptr;
+    _Atomic guint32* event_flag_ptr;
+    guint32 refcount;
+} GBinderFmqPriv;
+
+static inline GBinderFmqPriv* gbinder_fmq_cast(GBinderFmq* client)
+    { return G_CAST(client, GBinderFmqPriv, pub); }
+
+GBinderFmqGrantorDescriptor*
+gbinder_fmq_get_grantor_descriptor(
+    GBinderFmq* self,
+    gint index)
+{
+    return &((GBinderFmqGrantorDescriptor *)(
+        self->desc->grantors.data.ptr))[index];
+}
+
+gsize
+gbinder_fmq_available_to_read_bytes(
+    GBinderFmq* self,
+    gboolean contiguous)
+{
+    GBinderFmqPriv* priv = gbinder_fmq_cast(self);
+
+    guint64 read_ptr =
+        atomic_load_explicit(priv->read_ptr, memory_order_acquire);
+    gsize available_to_read =
+        atomic_load_explicit(priv->write_ptr, memory_order_acquire) - read_ptr;
+
+    if (contiguous) {
+        gsize size =
+            gbinder_fmq_get_grantor_descriptor(self, DATA_PTR_POS)->extent;
+        gsize read_offset = read_ptr % size;
+        /* The number of bytes that can be read contiguously from
+         * read_offset without wrapping around the ring buffer */
+        gsize available_to_read_contiguous = size - read_offset;
+        return available_to_read_contiguous < available_to_read ?
+            available_to_read_contiguous : available_to_read;
+    }
+    return available_to_read;
+}
+
+gsize
+gbinder_fmq_available_to_write_bytes(
+    GBinderFmq* self,
+    gboolean contiguous)
+{
+    GBinderFmqPriv* priv = gbinder_fmq_cast(self);
+
+    gsize available_to_write =
+        gbinder_fmq_get_grantor_descriptor(self, DATA_PTR_POS)->extent -
+        gbinder_fmq_available_to_read_bytes(self, FALSE);
+
+    if (contiguous) {
+        guint64 write_ptr = atomic_load_explicit(priv->write_ptr,
+            memory_order_relaxed);
+        gsize size =
+            gbinder_fmq_get_grantor_descriptor(self, DATA_PTR_POS)->extent;
+        gsize write_offset = write_ptr % size;
+        /* The number of bytes that can be written contiguously starting from
+         * write_offset without wrapping around the ring buffer */
+        gsize available_to_write_contiguous = size - write_offset;
+        return available_to_write_contiguous < available_to_write ?
+            available_to_write_contiguous : available_to_write;
+    }
+
+    return available_to_write;
+}
+
+static inline gsize align_to_word_boundary(gsize length) {
```

Comment:
This function looks functionally equivalent to `G_ALIGN8()` macro defined in `<gutil_macros.h>`

Reply:
Replaced with G_ALIGN8()